src/backend/cuda/convolve.hpp

```diff
@@ -11,12 +11,29 @@
 
 namespace cuda {
 
-template<typename T, typename accT, dim_t baseDim, bool expand>
-Array<T> convolve(Array<T> const& signal, Array<accT> const& filter,
+template <typename T, typename accT, dim_t baseDim, bool expand>
+Array<T> convolve(Array<T> const &signal, Array<accT> const &filter,
                   AF_BATCH_KIND kind);
 
-template<typename T, typename accT, bool expand>
-Array<T> convolve2(Array<T> const& signal, Array<accT> const& c_filter,
-                   Array<accT> const& r_filter);
+template <typename T, typename accT, bool expand>
+Array<T> convolve2(Array<T> const &signal, Array<accT> const &c_filter,
+                   Array<accT> const &r_filter);
 
-}  // namespace cuda
+template <typename T, typename accT>
+Array<T> convolve2(Array<T> const &signal, Array<accT> const &filter,
+                   const dim4 stride, const dim4 padding, const dim4 dilation);
+
+template <typename T, typename accT>
+Array<T> conv2DataGradient(const Array<T> &incoming_gradient,
+                           const Array<T> &original_signal,
+                           const Array<accT> &original_filter,
+                           const Array<T> &convolved_output, af::dim4 stride,
+                           af::dim4 padding, af::dim4 dilation);
+
+template <typename T, typename accT>
+Array<T> conv2FilterGradient(const Array<T> &incoming_gradient,
+                             const Array<T> &original_signal,
+                             const Array<accT> &original_filter,
+                             const Array<T> &convolved_output, af::dim4 stride,
+                             af::dim4 padding, af::dim4 dilation);
+}
```

Comment:
Just for consistency, can we add back `// namespace cuda` here?

Reply:
File is too short to have that.